src/main/scala/monadui/OutputAp.scala

```diff
@@ -0,0 +1,100 @@
+package monadui
+
+object LiftOutputTuples {
+  import scala.language.experimental.macros
+  import scala.reflect.macros.blackbox.Context
+
+  def apply[M[_],T](expr: M[T]): M[T] =  macro liftTuplesImpl[M[T],T]
+
+  def liftTuplesImpl[MT : c.WeakTypeTag, T : c.WeakTypeTag](c: Context)(expr: c.Tree) = {
+
+    import c.universe._
+
+    val flatMapName = TermName("flatMap")
+    val mapName = TermName("map")
+
+    object OpsUnwrap {
+      val tc = weakTypeOf[MT].typeConstructor
+      def unapply(v: c.Tree) = v match {
+        case v if v.tpe.typeConstructor == tc ⇒ Some(v)
+        case q"$ops($inner)" if inner.tpe.typeConstructor == tc ⇒ Some(inner)
+        case _ ⇒ None
+      }
+    }
+
+    class TupleLiftingTransformer extends Transformer {
+      // Perhaps Nested flatMap
+      private object PN {
+        def unapply(tree: Tree): Option[Tree] = tree match {
+          case Apply(TypeApply(Select(wr@OpsUnwrap(vm), TermName("flatMap")), tOuter),
+                     (oldClosure@Function(vValDef :: Nil,
+                     PN(Apply(TypeApply(Select(OpsUnwrap(wm), inner : TermName), tInner),
+                     (oldInnerClosure@Function(wValDef :: Nil, expr)) :: Nil)))) :: Nil)
+            if (inner == flatMapName || inner == mapName) ⇒
+
+            val ap2 = q"${wr.tpe.typeSymbol.companion}.ap2"
+
+            // Crucial part: ensure that v is never used as the qualifier
+            val vUsed = wm.find(vValDef.symbol == _.symbol)
```

Comment:
Beyond the proof of concept there are some edge cases consider around references in types. Need to traverse `sym.info` of `DefTrees` and/or the `.tpe` of all trees. 

``` 
LiftOutputTuples {
      for (_1 <- o1; _2 <- {def foo[T](t: T, x: () => Option[_1.type]):T = t; foo(o2, () => None)}; _3 <- o3; _4 <- o4; _5 <- o5; _6 <- o6; _7 <- o7; _8 <- o8; _9 <- o9; _10 <- o10) yield (_1 - _2 - _3 - _4 - _5 - _6 - _7 - _8 - _9 - _10)
    }
```

```
Information:(82, 24) Converted to OutputFlatMapDsl.RichOutput[(Int, (Int, (Int, (Int, (Int, (Int, (Int, (Int, (Int, Int)))))))))](RichOutput.ap2[Int, (Int, (Int, (Int, (Int, (Int, (Int, (Int, (Int, Int))))))))](OutputTest.this.o1, RichOutput.ap2[Int, (Int, (Int, (Int, (Int, (Int, (Int, (Int, Int)))))))](OutputTest.this.o2, RichOutput.ap2[Int, (Int, (Int, (Int, (Int, (Int, (Int, Int))))))](OutputTest.this.o3, RichOutput.ap2[Int, (Int, (Int, (Int, (Int, (Int, Int)))))](OutputTest.this.o4, RichOutput.ap2[Int, (Int, (Int, (Int, (Int, Int))))](OutputTest.this.o5, RichOutput.ap2[Int, (Int, (Int, (Int, Int)))](OutputTest.this.o6, RichOutput.ap2[Int, (Int, (Int, Int))](OutputTest.this.o7, RichOutput.ap2[Int, (Int, Int)](OutputTest.this.o8, RichOutput.ap2[Int, Int](OutputTest.this.o9, OutputTest.this.o10)))))))))).map[Int](((x$9: scala.Tuple2[Int, (Int, (Int, (Int, (Int, (Int, (Int, (Int, (Int, Int))))))))]) => {
  val _1: Int = x$9._1;
  val x$8: (Int, (Int, (Int, (Int, (Int, (Int, (Int, (Int, Int)))))))) = x$9._2;
  val _2: Int = x$8._1;
  val x$7: (Int, (Int, (Int, (Int, (Int, (Int, (Int, Int))))))) = x$8._2;
  val _3: Int = x$7._1;
  val x$6: (Int, (Int, (Int, (Int, (Int, (Int, Int)))))) = x$7._2;
  val _4: Int = x$6._1;
  val x$5: (Int, (Int, (Int, (Int, (Int, Int))))) = x$6._2;
  val _5: Int = x$5._1;
  val x$4: (Int, (Int, (Int, (Int, Int)))) = x$5._2;
  val _6: Int = x$4._1;
  val x$3: (Int, (Int, (Int, Int))) = x$4._2;
  val _7: Int = x$3._1;
  val x$2: (Int, (Int, Int)) = x$3._2;
  val _8: Int = x$2._1;
  val x$1: (Int, Int) = x$2._2;
  val _9: Int = x$1._1;
  val _10: Int = x$1._2;
  _1.-(_2).-(_3).-(_4).-(_5).-(_6).-(_7).-(_8).-(_9).-(_10)
}))
      LiftOutputTuples {
Error:(88, 62) type mismatch;
 found   : _1.type (with underlying type Int)
 required: AnyRef
      for (_1 <- o1; _2 <- {def foo[T](t: T, x: () => Option[_1.type]):T = t; foo(o2, () => None)}; _3 <- o3; _4 <- o4; _5 <- o5; _6 <- o6; _7 <- o7; _8 <- o8; _9 <- o9; _10 <- o10) yield (_1 - _2 - _3 - _4 - _5 - _6 - _7 - _8 - _9 - _10)
```

Reply:
Good point, though I'd be surprised if this were the last unaddressed edge case.